### PR TITLE
Re-implement Wavefront Histogram

### PR DESCRIPTION
**Pull request by haosong**
Re-implement `WavefrontHistogramImpl` as follows:

1. Created a new `ThreadMinuteBin` replacing `MinuteBin` class, which contains minute bins from all threads in one minute. 
2. Changed `globalHistogramBinsList` as `ConcurrentLinkedDeque<ThreadMinuteBin>`. Removed the usage of `WeakReference` and `ThreadLocal histogramBinsList`.
3. Used one local variable `ThreadMinuteBin currentMinuteBin` for `update()` funcs to feed data in.
4. Every time when `getCurrentBin()` or `getGlobalHistogramBinsList()` is called, a synchronized function `flushCurrentBin()` will be called to check whether the `currMinuteMillis` is the same as the timestamp inside `currentMinuteBin`, if not flush that bin into `globalHistogramBinsList` and reset `currentMinuteBin`.

All the producers will only produce data into the local variable `currentMinuteBin`. All the consumers will only consume data from `globalHistogramBinsList`. `flushCurrentBin()` is responsible for appending `currentMinuteBin` to `globalHistogramBinsList`.

### CODE REVIEW

File: src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java [hunk 1]
```diff
@@ -377,20 +308,66 @@ public Distribution(long timestamp, List<Pair<Double, Integer>> centroids) {
...
+      this.perThreadDist = new ConcurrentHashMap<>(threadMinuteBin.perThreadDist);
+      this.minuteMillis = threadMinuteBin.minuteMillis;
+    }
+
+    /**
+     * Helper to retrieve the thread-local distribution in one given minute.
+     */
+    TDigest getDistByThreadId(long threadId) {
+      // Create new Digest for new thread.
+      return perThreadDist.computeIfAbsent(threadId, digest -> new AVLTreeDigest(ACCURACY));
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java]: should be threadId, key -> new AVLTreeDigest(ACCURACY)
as digest is not long, 
timestampMillis is the key here that is of type long.

File: src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java [hunk 2]
```diff
@@ -42,45 +40,26 @@
...
-        ConcurrentLinkedDeque<MinuteBin> sharedBinsInstance = new ConcurrentLinkedDeque<>();
-        try {
-          writeLock.lock();
-          globalHistogramBinsList.add(new WeakReference<>(sharedBinsInstance));
-        } finally {
-          writeLock.unlock();
-        }
-        return sharedBinsInstance;
-  });
+  private ThreadMinuteBin currentMinuteBin;
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java]: Can you add a comment as to why this is not final and how updating this reference is done in thread safe manner below.

File: src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java [hunk 3]
```diff
@@ -266,19 +178,38 @@ private long currentMinuteMillis() {
...
-        sharedBinsInstance.getLast().minuteMillis != currMinuteMillis) {
-      sharedBinsInstance.add(new MinuteBin(ACCURACY, currMinuteMillis));
-      if (sharedBinsInstance.size() > MAX_BINS) {
-        sharedBinsInstance.removeFirst();
+    return flushCurrentBin(currMinuteMillis);
+  }
+
+  private ThreadMinuteBin flushCurrentBin(long currMinuteMillis) {
+    if (this.currentMinuteBin.minuteMillis == currMinuteMillis) return this.currentMinuteBin;
+    synchronized(this) {
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java]: nit: should be
synchronized (this) {
Also might be better to synchronize on 'globalHistogramBinsList'

File: src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java [hunk 4]
```diff
@@ -266,19 +178,38 @@ private long currentMinuteMillis() {
...
+      if (this.currentMinuteBin.minuteMillis != currMinuteMillis) {
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java]: Add a comment as to why double check inside synchronized (this) block is important to do here.
if (this.currentMinuteBin.minuteMillis != currMinuteMillis) { ...

File: src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java [hunk 5]
```diff
@@ -266,19 +178,38 @@ private long currentMinuteMillis() {
...
     long currMinuteMillis = currentMinuteMillis();
-    if (sharedBinsInstance.isEmpty() ||
-        sharedBinsInstance.getLast().minuteMillis != currMinuteMillis) {
-      sharedBinsInstance.add(new MinuteBin(ACCURACY, currMinuteMillis));
-      if (sharedBinsInstance.size() > MAX_BINS) {
-        sharedBinsInstance.removeFirst();
+    return flushCurrentBin(currMinuteMillis);
+  }
+
+  private ThreadMinuteBin flushCurrentBin(long currMinuteMillis) {
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java]: getOrUpdateCurrentBin is a better method name which conveys the meaning better.
We are not really flushing the bin here so the name does not seem correct to me.

File: src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java [hunk 6]
```diff
@@ -42,45 +40,26 @@
...
-  private final List<WeakReference<ConcurrentLinkedDeque<MinuteBin>>> globalHistogramBinsList =
-      new ArrayList<>();
-
-  private final StampedLock stampedLock = new StampedLock();
-  // Protects read access to globalHistogramBinsList
-  private final Lock readLock = stampedLock.asReadLock();
-
-  // Protects write access to globalHistogramBinsList
-  private final Lock writeLock = stampedLock.asWriteLock();
+  private final ConcurrentLinkedDeque<ThreadMinuteBin> globalHistogramBinsList = new ConcurrentLinkedDeque<>();
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java]: Please rename this to priorMinuteBinsList
this includes all ThreadMinuteBin prior to current bin. Please update the doc accordingly.
"Global concurrent list of ThreadMinuteBin." is not good enough.

**Comment by panghy** [comment]
Quick thing, can we include a benchmark, both single threaded and multi
threaded for the change? It can be as simple as an @Ignored unit test.

**Comment by song-hao** [comment]
See benchmark here: https://github.com/wavefrontHQ/wavefront-java-sdk/pull/26

**Comment by panghy** [comment]
@haosong this is still open? please address the conflicts or close

**Comment by haosong** [comment]
Close the PR for now

**Pull request closed without merging**